pretyping/evarsolve.ml

```diff
@@ -1246,33 +1246,51 @@ let check_evar_instance evd evk1 body conv_algo =
   | Success evd -> evd
   | UnifFailure _ -> raise (IllTypedInstance (evenv,ty,EConstr.of_constr evi.evar_concl))
 
-let update_evar_source ev1 ev2 evd =
-  let loc, evs2 = evar_source ev2 evd in
-  match evs2 with
-  | (Evar_kinds.QuestionMark _ | Evar_kinds.ImplicitArg (_, _, false)) ->
-     let evi = Evd.find evd ev1 in
-     Evd.add evd ev1 {evi with evar_source = loc, evs2}
-  | _ -> evd
-  
+let update_evar_info ev1 ev2 evd =
+  let evi = find evd ev2 in
+  let loc, evs2 = evi.evar_source in
+  let evi = Evd.find evd ev1 in
+  let evi' =
+    match evs2 with
+    | (Evar_kinds.QuestionMark _ | Evar_kinds.ImplicitArg (_, _, false)) ->
+       {evi with evar_source = loc, evs2}
+    | _ -> evi
+  in
+  let evi' = { evi with evar_extra = Evd.Store.merge evi.evar_extra evi'.evar_extra } in
+  Evd.add evd ev1 evi'
+
 let solve_evar_evar_l2r force f g env evd aliases pbty ev1 (evk2,_ as ev2) =
   try
     let evd,body = project_evar_on_evar force g env evd aliases 0 pbty ev1 ev2 in
     let evd' = Evd.define evk2 (EConstr.Unsafe.to_constr body) evd in
-    let evd' = update_evar_source (fst (destEvar evd body)) evk2 evd' in
+    let evd' = update_evar_info (fst (destEvar evd body)) evk2 evd' in
       check_evar_instance evd' evk2 body g
   with EvarSolvedOnTheFly (evd,c) ->
     f env evd pbty ev2 c
 
 let opp_problem = function None -> None | Some b -> Some (not b)
 
+(** Support for program obligations: during merges of evars, the obligation flag is kept *)
+let merge_obligation xy =
```

Comment:
That's the default, right? Why not rename the default as `don't_care` (as in « don't care non-determinism »), or `choose_one` or `choose_first` or something like that and commit to this particular semantics (so that choosing default doesn't feel like shrugging off and pick a random semantics), and just use it here as well?